plugins/out_skywalking/skywalking.c

```diff
@@ -0,0 +1,387 @@
+    } else {
+        io_flags = FLB_IO_TCP;
+        ctx->http_scheme = flb_sds_create("http://");
+    }
+
+    /* setup host and port */
+    ctx->host = ins->host.name;
+    ctx->port = ins->host.port;
+    flb_plg_debug(ctx->ins, "OAP address: %s:%d", ctx->host, ctx->port);
+
+    /* configure upstream instance */
+    ctx->u = flb_upstream_create(config, ctx->host, ctx->port, io_flags, ins->tls);
+    if (!ctx->u) {
+        flb_plg_error(ctx->ins, "failed to create upstream context");
+        flb_free(ctx);
+        return -1;
+    }
+
+    flb_output_upstream_set(ctx->u, ins);
+
+    /* Set the plugin context */
+    flb_output_set_context(ins, ctx);
+    flb_output_set_http_debug_callbacks(ins);
+
+    return 0;
+}
+
+static int64_t timestamp_format(const struct flb_time* tms) {
+  int64_t timestamp = 0;
+
+  /* Format the time, use milliseconds precision not nanoseconds */
+  timestamp = tms->tm.tv_sec * 1000;
+  timestamp += tms->tm.tv_nsec / 1000000;
+
+  /* round up if necessary */
+  if (tms->tm.tv_nsec % 1000000 >= 500000) {
+      ++timestamp;
+  }
+  return timestamp;
+}
+
+static void sw_msgpack_pack_kv_str(msgpack_packer* pk, const char* key,
+                                   size_t key_len, const char *value, size_t value_len) {
+    msgpack_pack_str(pk, key_len);
+    msgpack_pack_str_body(pk, key, key_len);
+    msgpack_pack_str(pk, value_len);
+    msgpack_pack_str_body(pk, value, value_len);
+}
+
+static void sw_msgpack_pack_kv_int64_t(msgpack_packer* pk, const char* key,
+                                       size_t key_len, int64_t value) {
+    msgpack_pack_str(pk, key_len);
+    msgpack_pack_str_body(pk, key, key_len);
+    msgpack_pack_int64(pk, value);
+}
+
+static void sw_msgpack_pack_log_body(msgpack_packer* pk,
+                                     msgpack_object* obj, size_t obj_size) {
+    int i, j = 0;
+    int log_entry_num = 0;
+    msgpack_sbuffer sbuf;
+    msgpack_packer body_pk;
+    msgpack_object key;
+    msgpack_object value;
+    flb_sds_t out_body_str;
+    size_t out_body_str_len;
+    int* valid_log_entry = (int*)flb_malloc(obj_size * sizeof(int));
+
+    msgpack_sbuffer_init(&sbuf);
+    msgpack_packer_init(&body_pk, &sbuf, msgpack_sbuffer_write);
+
+    for (i = 0; i < obj_size; ++i) {
+        key = obj->via.map.ptr[i].key;
+        value = obj->via.map.ptr[i].val;
+
+        if (key.type != MSGPACK_OBJECT_STR || value.type != MSGPACK_OBJECT_STR) {
+            continue;
+        }
+
+        valid_log_entry[j] = i;
+        ++j;
+        ++log_entry_num;
+    }
+
+    msgpack_pack_map(&body_pk, log_entry_num);
+
+    for (i = 0; i < log_entry_num; ++i) {
+        key = obj->via.map.ptr[valid_log_entry[i]].key;
+        value = obj->via.map.ptr[valid_log_entry[i]].val;
+        sw_msgpack_pack_kv_str(&body_pk, key.via.str.ptr, key.via.str.size, value.via.str.ptr, value.via.str.size);
+    }
+
+    out_body_str = flb_msgpack_raw_to_json_sds(sbuf.data, sbuf.size);
```

Comment:
if out_body_str gets NULL the next line will crash it

Reply:
added NULL check